### DISCUSSION TITLE
MQTT: Is there a smarter way?

### DISCUSSION THREAD

**Discussion by UBWH** [arendst/Tasmota#20750] [General]
I have been working on a project that polls a device every second using Modbus.

I have it working using 2 methods:

BERRY
- Modbus Driver: 2 GPIO pins set to `ModBr TX`, `ModBr Rx`, so using the built-in Tasmota Modbus drivers
- Every second: tasmota.cmd("'modbussend {..."), and then capture the data with a Rule that parses the ModbusReceived messages
- Side effect is MQTT message ("ModbusSend": "Done") issued every second.
- BERRY coding is simpler with more flexibility than SCRIPT
- The 'problem' here is that I am using a high-level command, to do the low-level device (sensor) polling work

SCRIPT
- Modbus Driver: Using the [Smart Meter Interface](https://tasmota.github.io/docs/Smart-Meter-Interface/)
- Sends/Parses Modbus messages & replies every second. 
- Does not use the Tasmota modbussend command, so no excess MQTT traffic
- But ... found heap use was higher than BERRY method, and that eventually caused me to abandon the SCRIPT method

Thanks to all the help I have received for many sources to help get as far as I have so far.

Here is my question:
Is there any way I can make a BERRY Driver to call the low-level functions in [xdrv_63_modbus_bridge.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_63_modbus_bridge.ino), as a way of bypassing the higher level ` tasmota.cmd("'modbussend {...")` ?

**Comment by sfromis** [comment]
There is no feature to call native driver functions from Berry, without the driver already having extra support for such. Maybe it would be time for "someone" to create a Berry/Modbus integration.

Not clear what you mean by "using MQTT" when you seem to be wanting to not have MQTT activity. That may depend on what you're doing with the captured `ModbusReceived` data, and generally on how the broader data flow is. Alternative solutions would depend on details, like if you'd be able to use another protocol (Berry has several options there), and turn off MQTT.

**Comment by UBWH** [comment]
"Using MQTT": Typo. Fixed. Now says "Using Modbus"

Thanks for the reply. I thought that was the case .. but thought I would ask anyway.
